File: Convert Sorted Array to Binary Search Tree.py
# Leetcode Problem 108: Convert Sorted Array to Binary Search Tree
# PYTHON CODE
class Solution:
    def sortedArrayToBST(self, nums):
        
        def build(left, right):
            
            # No elements
            if left > right:
                return None

            # Middle element becomes root
            mid = (left + right) // 2

            root = TreeNode(nums[mid])

            # Build left subtree
            root.left = build(left, mid - 1)

            # Build right subtree
            root.right = build(mid + 1, right)

            return root

        return build(0, len(nums) - 1)

File: Convert Sorted Array to Binary Search Tree.java
// Leetcode Problem 108: Convert Sorted Array to Binary Search Tree 
// JAVA CODE
class Solution {

    public TreeNode sortedArrayToBST(int[] nums) {
        return buildTree(nums, 0, nums.length - 1);
    }

    private TreeNode buildTree(int[] nums, int left, int right) {

        // No elements
        if (left > right) {
            return null;
        }

        // Find middle element
        int mid = left + (right - left) / 2;

        // Middle element becomes root
        TreeNode root = new TreeNode(nums[mid]);

        // Build left subtree
        root.left = buildTree(nums, left, mid - 1);

        // Build right subtree
        root.right = buildTree(nums, mid + 1, right);

        return root;
    }
}